src/PulseAudioManager.vala

```diff
@@ -245,143 +402,276 @@ public class Sound.PulseAudioManager : GLib.Object {
      * Retrieve object informations
      */
 
-    private void source_info_callback (PulseAudio.Context c, PulseAudio.SourceInfo? i, int eol) {
-        if (i == null)
+    private void source_info_callback (PulseAudio.Context c, PulseAudio.SourceInfo? source, int eol) {
+        if (source == null) {
             return;
+        }
 
         // completely ignore monitors, they're not real sources
-        if (i.monitor_of_sink != PulseAudio.INVALID_INDEX) {
+        if (source.monitor_of_sink != PulseAudio.INVALID_INDEX) {
             return;
         }
+        debug ("source info update");
+        debug ("  source: %s (%s)", source.description, source.name);
+        debug ("    card: %u", source.card);
 
-        Device device = null;
-        bool is_new = !input_devices.has_key (i.index);
-        if (is_new) {
-            device = new Device (i.index);
-        } else {
-            device = input_devices.get (i.index);
+        if (source.name == "auto_null") {
+            return;
         }
 
-        device.input = true;
-        device.name = i.name;
-        device.display_name = i.description;
-        device.is_muted = (i.mute != 0);
-        device.cvolume = i.volume;
-        device.channel_map = i.channel_map;
-        device.balance = i.volume.get_balance (i.channel_map);
-        device.volume_operations.foreach ((operation) => {
-            if (operation.get_state () != PulseAudio.Operation.State.RUNNING) {
-                device.volume_operations.remove (operation);
-            }
-
-            return GLib.Source.CONTINUE;
-        });
+        foreach (var port in source.ports) {
```

Comment:
This loop shouldn't probably be in there when not in debug mode.

Reply:
Sorry, I was compiling the sound indicator, missing the many errors and warnings. Fixes changes: https://github.com/elementary/switchboard-plug-sound/compare/0e19945cd64072b255b5065a35049ed2c8e69b5d...change-devices-approach